File: linked_list.mjs
class LinkedList{
    constructor(head = null){
        this.head = head
    }


    append(value){
        
        //add node to end of linked list
        let newNode = new node(value)
        let start = this.head

        //if list is empty
        if(start == null){
            this.head = newNode
            return
        }
        while(start.nextNode != null){
            start = start.nextNode
        }
        start.nextNode = newNode
    }

    prepend(value){
        //add node to start of linked list

        let newNode = new node(value)

        //if list is empty
        if(this.head == null){
            this.head = newNode
            return
        }
        
        newNode.nextNode = this.head
        this.head = newNode
    }

    size(){
        //return number of nodes in linked list

        let size = 0
        let start = this.head
        
        while(start != null){
            size += 1
            start = start.nextNode
        }

        return size
    }

    head1(){
        //returns first node of linked list

        return (this.head)
    }

    tail(){
        //returns last node of linked list

        let start = this.head

        while(start.nextNode != null){
            start = start.nextNode
        }
        return start
    }
    at(index){
        //returns node of linked list at given index

        let start = this.head
        let i = 0
        while(i < index){
            i++
            start = start.nextNode
        }
        return start
    }

    pop(){
        //remove last node of linked list   

        let start = this.head

        if(this.size() == 1){
            start.value = null
            start.nextNode = null
            return
        }
        while(start.nextNode.nextNode != null){
            start = start.nextNode
        }
        start.nextNode = null
    }

    contains(value){
        //checks if linked list contains given value

        let start = this.head
        while(start){
            if(start.value == value){
                return true
            }
            else{
                start = start.nextNode
            }
        }
        return false
    }
    find(value){
        //finds index at which value is located at in linked list

        let start = this.head
        let length = this.size()
        for(let i=0; i < length; i++){
            if(start.value == value){
                return i
            }
            start = start.nextNode
        }
        return null
    }

    toString(){
        //represents linked list in form of a neat string

        let start = this.head
        let str1 = ''
        while(start){
            str1 += '( ' + (start.value).toString() + ' ) -> '
            start = start.nextNode
        }
        str1 += 'null'
        return str1
    }

    insertAt(value, index){
        //insert node of value at index in the linked list

        let newNode = new node(value)

        let start = this.head
        let length = this.size() - 1

        let prevNextNode = null
        let aheadNextNode = null

        for(let i=0; i <= length; i++){
            if(i+1 == index){
                prevNextNode = start
            }
            if(i == index){
                aheadNextNode = start
                break
            }
            start = start.nextNode
        }
        prevNextNode.nextNode = newNode
        newNode.nextNode = aheadNextNode
    }

    removeAt(index){
        //removes node from certain index from linked list

        let start = this.head
        let length = this.size() - 1

        let prevNextNode = null
        let aheadNextNode = null

        for(let i=0; i <= length; i++){
            if(i+1 == index){
                prevNextNode = start
            }
            if(i-1 == index){
                aheadNextNode = start
                break
            }
            start = start.nextNode
        }
        prevNextNode.nextNode = aheadNextNode

    }
}

class node{
    constructor(value = null, nextNode = null){
        this.value = value
        this.nextNode = nextNode
    }
}

export {LinkedList, node}

//test cases
// let list = new LinkedList()

// list.append(8)
// list.append(10)
// list.append(12)
// list.append(14)
// list.append(16)
// console.log(list.size())
// console.log(list.head1())
// console.log(list.tail())
// console.log('index', list.at(0))

// console.log(list.contains(12))
// console.log(list.find(8))

// console.log(list.toString())
// list.insertAt('Imposter',3)
// console.log(list.toString())
// list.removeAt(3)
// console.log(list.toString())
// console.log(list)

File: linked_list.c
#include <stdio.h>
#include <stdlib.h>


// Define the structure for a node
struct Node {
    int data;
    struct Node* next;
};

struct LinkedList {
    int size;
};

struct LinkedList list = {0};

struct Node* Head = NULL;

void append(int data); //insertion at end O(n)
void endRemoval(void);   //remove last node O(n)
void prepend(int data); //append at start  O(1)
void insertAt(int data, int index); //insert at given index  O(n)
void removeAt(int index); //remove at given idnex  O(n)
void freeList(void);

int main() {
    // Create the first node (head)
    append(10);
    append(20);
    append(30);
    append(40);
    endRemoval();
    //endRemoval();
    //prepend(10);
    //insertAt(25, 2);
    //freeList();


    // Traversing the linked list
    struct Node* current = Head;
    printf("Linked List: ");
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
    printf("Size of Linked List: %d", list.size);
    return 0;
}

void append(int data){

    struct Node* new_node = malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->next = NULL;

    if(Head == NULL){
        Head = new_node;
        list.size += 1;
    }
    else{
        struct Node* current = Head;
        while(current->next != NULL){
            current = current -> next;
        }
        current -> next = new_node;
        list.size += 1;
    }
}

void endRemoval(void){

    if(Head == NULL){
        printf("Linked List is currently empty\n");
        return;
    }
    if (Head->next == NULL) { // Case where there's only one node
        free(Head);
        Head = NULL;
        list.size -= 1;
        return;
    }
    struct Node* current = Head;
    while(current->next->next != NULL){
        current = current->next;
    }
    struct Node* lastNode = current->next;
    current->next = NULL;
    free(lastNode);
    list.size -= 1;
}

void prepend(int data){
    struct Node* new_node = malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->next = NULL;

    new_node->next = Head;
    Head = new_node;
    list.size += 1;
}

void insertAt(int data, int index){

    if (index == 0) {
        prepend(data);
        return;
    }
    if (index == list.size) {
        append(data);
        return;
    }
    if (index < 0 || index > list.size) {
        printf("Enter Valid index!\n");
        return;
    }
    struct Node* new_node = malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->next = NULL;
    
    struct Node* current = Head;
    int i = 0;
    while(i < index - 1){
        current = current -> next;
        i++;
    }
    struct Node* previousNode = current;
    struct Node* nextNode = current -> next;

    previousNode->next = new_node;
    new_node->next = nextNode; 
    list.size += 1;
}

void removeAt(int index){
    if(index < 0 || index >= list.size){
        printf("Enter Valid index!\n");
        return;
    }

    if(index == 0){
        struct Node* current = Head;
        Head = Head->next;
        free(current);
        list.size -= 1;
        return;
    }
    if(index == list.size - 1){
        struct Node* current = Head;
        while(current -> next -> next != NULL){
            current = current -> next;
        }
        struct Node* lastNode = current -> next-> next;
        current->next = NULL;
        free(lastNode);
        list.size -= 1;
        return;
    }
    int i = 0;
    struct Node* current = Head;
    while(i < index - 1){
        current = current -> next;
        i++;
    }
    struct Node* nextNode = current -> next -> next;
    free(current-> next);
    current-> next = nextNode;
    list.size -= 1;
}

void freeList(void) {
    struct Node* current = Head;
    struct Node* nextNode;
    while (current != NULL) {
        nextNode = current->next;
        free(current);
        current = nextNode;
    }
    Head = NULL;
    list.size = 0;
}
